pkg/blobstore/cassandra/cassandra_blob_access.go

```diff
@@ -0,0 +1,673 @@
+		Name:      "cassandra_total_write_duration_in_seconds",
+		Help:      "Total amount of time in seconds it takes to write an entire bytestream to Cassandra",
+		Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+	})
+
+	lastAccessUpdatesFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "buildbarn",
+		Subsystem: "storage",
+		Name:      "cassandra_last_access_failure_count",
+		Help:      "Count of the number of times the `last_access` field could not be updated",
+	})
+)
+
+type job struct {
+	task func()
+}
+
+type cassandraBlobAccess struct {
+	capabilitiesProvider     capabilities.Provider
+	readBufferFactory        blobstore.ReadBufferFactory
+	segmentSize              int32
+	universalInstanceName    string
+	session                  *gocql.Session
+	lastAccessUpdateInterval time.Duration
+	lastAccessUpdate         chan<- job
+	tables                   *tables
+}
+
+// NewCassandraBlobAccess provides an implementation of storage backed by Cassandra.
+func NewCassandraBlobAccess(capabilitiesProvider capabilities.Provider, readBufferFactory blobstore.ReadBufferFactory, tlsConfig *tls.Config, port, protoVersion int32, username, password string, clusterHosts []string, keyspace, tablePrefix string, segmentSizeInBytes int32, preferredDc string, lastAccessUpdateInterval time.Duration, universalInstanceName string) blobstore.BlobAccess {
+	cassandraBlobStorePrometheusMetrics.Do(func() {
+		prometheus.MustRegister(segmentWriteHist)
+		prometheus.MustRegister(segmentReadHist)
+		prometheus.MustRegister(totalWritesHist)
+		prometheus.MustRegister(lastAccessUpdatesFailures)
+	})
+
+	cluster := gocql.NewCluster(clusterHosts...)
+	cluster.Keyspace = keyspace
+	cluster.Consistency = writeConsistency
+	if preferredDc != "" {
+		cluster.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(gocql.DCAwareRoundRobinPolicy(preferredDc))
+	}
+
+	if port != 0 {
+		cluster.Port = int(port)
+	}
+
+	if protoVersion != 0 {
+		cluster.ProtoVersion = int(protoVersion)
+	}
+
+	if username != "" {
+		cluster.Authenticator = gocql.PasswordAuthenticator{
+			Username: username,
+			Password: password,
+		}
+	}
+
+	if tlsConfig != nil {
+		cluster.SslOpts = &gocql.SslOptions{
+			Config: tlsConfig,
+		}
+	} else {
+		log.Printf("Not setting up TLS config")
+	}
+
+	if tablePrefix == "" {
+		log.Printf("No table prefix set. Cowardly refusing to proceed.")
+		return blobstore.NewErrorBlobAccess(errors.New("table prefix must be set"))
+	}
+	log.Printf("Table prefix: %s", tablePrefix)
+
+	log.Printf("Creating session in keyspace '%s' for %v", keyspace, clusterHosts)
+
+	session, err := cluster.CreateSession()
+	if err != nil {
+		log.Printf("Unable to create session: %v", err)
+		return blobstore.NewErrorBlobAccess(err)
+	}
+
+	// Create the worker pool for updating the last access times
+	jobs := make(chan job, lastAccessUpdateWorkerCount)
+	for i := 0; i < lastAccessUpdateWorkerCount; i++ {
+		go lastUpdateWorker(jobs)
+	}
```

Comment:
As discussed, please move this into a separate function, as it prevents cassandraBlobAccess from being unit tested properly. Also remove the `log.Printf()` calls, as there is no need to repeat whatever is specified in config. Because once we start to tolerate this, the question becomes: "Why do we print the value of this one configuration option, but not of this other one?" Also get rid of the calls to NewErrorBlobAccess. Any misconfigurations should lead to a fatal startup error.